File: backend/app/rag/chunking.py
"""Section-aware chunking: 500–800 tokens with ~100-token overlap."""

from __future__ import annotations

import re
from dataclasses import dataclass

from app.rag.cleanup import cleanup_text
from app.rag.tokenize import estimate_tokens

_HEADING_MD = re.compile(r"^(#{1,3})\s+(.+)$")
_HEADING_PLAIN = re.compile(r"^([A-Z][A-Za-z0-9 /&-]{3,80})$")


@dataclass
class ChunkDraft:
    ordinal: int
    section: str
    page: int | None
    text: str
    token_count: int


@dataclass
class ChunkOptions:
    min_tokens: int = 500
    max_tokens: int = 800
    overlap_tokens: int = 100


DEFAULTS = ChunkOptions()


def extract_sections(text: str) -> list[dict[str, str]]:
    lines = text.replace("\r\n", "\n").split("\n")
    sections: list[dict[str, str]] = []
    current = "Introduction"
    buf: list[str] = []

    def flush() -> None:
        nonlocal buf
        body = "\n".join(buf).strip()
        if body:
            sections.append({"section": current, "body": body})
        buf = []

    for line in lines:
        heading = _HEADING_MD.match(line) or _HEADING_PLAIN.match(line)
        if heading and len(line.strip()) < 90:
            flush()
            current = heading.group(2) if heading.lastindex and heading.lastindex >= 2 else heading.group(1)
            current = re.sub(r"^#+\s*", "", current).strip()
        else:
            buf.append(line)
    flush()
    return sections if sections else [{"section": "Body", "body": text.strip()}]


def overlap_tail(text: str, overlap_tokens: int) -> str:
    words = text.split()
    if len(words) <= overlap_tokens:
        return text
    return " ".join(words[-overlap_tokens:])


def chunk_document(text: str, options: ChunkOptions | None = None) -> list[ChunkDraft]:
    opts = options or DEFAULTS
    cleaned = cleanup_text(text)
    sections = extract_sections(cleaned)
    chunks: list[ChunkDraft] = []
    ordinal = 0

    def push(raw: str, section: str) -> None:
        nonlocal ordinal
        trimmed = raw.strip()
        if not trimmed:
            return
        ordinal += 1
        chunks.append(
            ChunkDraft(
                ordinal=ordinal,
                section=section,
                page=max(1, (ordinal + 1) // 2),
                text=trimmed,
                token_count=estimate_tokens(trimmed),
            )
        )

    for section in sections:
        paragraphs = [p.strip() for p in re.split(r"\n{2,}", section["body"]) if p.strip()]
        buffer = ""
        for para in paragraphs:
            nxt = f"{buffer}\n\n{para}" if buffer else para
            if estimate_tokens(nxt) > opts.max_tokens and buffer:
                push(buffer, section["section"])
                overlap = overlap_tail(buffer, opts.overlap_tokens)
                buffer = f"{overlap}\n\n{para}" if overlap else para
            else:
                buffer = nxt
        if buffer:
            if estimate_tokens(buffer) < opts.min_tokens and chunks:
                last = chunks[-1]
                if last.section == section["section"] and estimate_tokens(last.text + buffer) <= opts.max_tokens + 80:
                    last.text = f"{last.text}\n\n{buffer}".strip()
                    last.token_count = estimate_tokens(last.text)
                else:
                    push(buffer, section["section"])
            else:
                push(buffer, section["section"])

    if chunks:
        return chunks
    return [
        ChunkDraft(
            ordinal=1,
            section="Body",
            page=1,
            text=cleaned,
            token_count=estimate_tokens(cleaned),
        )
    ]

File: src/lib/pilot/rag/chunking.ts
import { estimateTokens } from "./tokenize";

export interface ChunkDraft {
  ordinal: number;
  section: string;
  page: number | null;
  text: string;
  tokenCount: number;
}

export interface ChunkOptions {
  minTokens?: number;
  maxTokens?: number;
  overlapTokens?: number;
}

const DEFAULTS: Required<ChunkOptions> = {
  minTokens: 500,
  maxTokens: 800,
  overlapTokens: 100,
};

export function extractSections(text: string): { section: string; body: string }[] {
  const lines = text.replace(/\r\n/g, "\n").split("\n");
  const sections: { section: string; body: string }[] = [];
  let current = "Introduction";
  let buf: string[] = [];
  const flush = () => {
    const body = buf.join("\n").trim();
    if (body) sections.push({ section: current, body });
    buf = [];
  };
  for (const line of lines) {
    const heading = /^(#{1,3})\s+(.+)$/.exec(line) || /^([A-Z][A-Za-z0-9 /&-]{3,80})$/.exec(line);
    if (heading && line.trim().length < 90) {
      flush();
      current = (heading[2] ?? heading[1]).replace(/^#+\s*/, "").trim();
    } else {
      buf.push(line);
    }
  }
  flush();
  return sections.length ? sections : [{ section: "Body", body: text.trim() }];
}

export function chunkDocument(text: string, options: ChunkOptions = {}): ChunkDraft[] {
  const opts = { ...DEFAULTS, ...options };
  const cleaned = cleanupText(text);
  const sections = extractSections(cleaned);
  const chunks: ChunkDraft[] = [];
  let ordinal = 0;

  for (const section of sections) {
    const paragraphs = section.body.split(/\n{2,}/).map((p) => p.trim()).filter(Boolean);
    let buffer = "";
    const push = (raw: string) => {
      const trimmed = raw.trim();
      if (!trimmed) return;
      ordinal += 1;
      chunks.push({
        ordinal,
        section: section.section,
        page: Math.max(1, Math.ceil(ordinal / 2)),
        text: trimmed,
        tokenCount: estimateTokens(trimmed),
      });
    };

    for (const para of paragraphs) {
      const next = buffer ? `${buffer}\n\n${para}` : para;
      if (estimateTokens(next) > opts.maxTokens && buffer) {
        push(buffer);
        const overlap = overlapTail(buffer, opts.overlapTokens);
        buffer = overlap ? `${overlap}\n\n${para}` : para;
      } else {
        buffer = next;
      }
    }
    if (buffer) {
      if (estimateTokens(buffer) < opts.minTokens && chunks.length) {
        const last = chunks[chunks.length - 1];
        if (last.section === section.section && estimateTokens(last.text + buffer) <= opts.maxTokens + 80) {
          last.text = `${last.text}\n\n${buffer}`.trim();
          last.tokenCount = estimateTokens(last.text);
        } else {
          push(buffer);
        }
      } else {
        push(buffer);
      }
    }
  }

  return chunks.length ? chunks : [{
    ordinal: 1,
    section: "Body",
    page: 1,
    text: cleaned,
    tokenCount: estimateTokens(cleaned),
  }];
}

export function cleanupText(text: string): string {
  return text
    .split("\u0000")
    .join("")
    .replace(/\r\n/g, "\n")
    .replace(/[ \t]+\n/g, "\n")
    .replace(/\n{3,}/g, "\n\n")
    .replace(/<!--[\s\S]*?-->/g, "")
    .trim();
}

function overlapTail(text: string, overlapTokens: number): string {
  const words = text.split(/\s+/);
  if (words.length <= overlapTokens) return text;
  return words.slice(-overlapTokens).join(" ");
}
